src/modules/ekf2/EKF2Selector.cpp

```diff
@@ -241,43 +242,77 @@ void EKF2Selector::Run()
 		updateParams();
 	}
 
-	// update combined test ratio for all estimators
-	updateErrorScores();
-
 	_sensors_status_imu.update();
 	const sensors_status_imu_s &sensors_status_imu = _sensors_status_imu.get();
 
+	_gyro_fault_detected = false;
+	uint32_t faulty_gyro_id = 0;
 	if (_selected_instance != UINT8_MAX) {
 		static constexpr uint8_t IMU_STATUS_SIZE = (sizeof(sensors_status_imu.gyro_inconsistency_rad_s) / sizeof(
 					sensors_status_imu.gyro_inconsistency_rad_s[0]));
-
+		static constexpr uint8_t IMU_INDEX_LIMIT = math::min(IMU_STATUS_SIZE, MAX_INSTANCES);
+		const float angle_rate_threshold = math::radians(_param_ekf2_sel_gyr_rate.get());
+		const float angle_threshold = math::radians(_param_ekf2_sel_gyr_angle.get());
+		const float time_step_s = fminf(1E-6f * (float)hrt_elapsed_time(&_last_update_us), 1.0f);
 		if (sensors_status_imu.gyro_device_id_primary == _instance[_selected_instance].estimator_status.gyro_device_id) {
-			for (unsigned i = 0; i < IMU_STATUS_SIZE; i++) {
+			// accumulate excess gyro error
+			uint8_t n_gyros = 0;
+			uint8_t n_exceedances = 0;
+			for (unsigned i = 0; i < IMU_INDEX_LIMIT; i++) {
 				if (sensors_status_imu.gyro_device_ids[i] != 0) {
-					// TODO: compare gyro inconsistency
-
-
-
-
+					n_gyros++;
+					if (sensors_status_imu.gyro_device_ids[i] == sensors_status_imu.gyro_device_id_primary) {
+						_accumulated_gyro_error[i] = 0.0f;
+					} else {
+						_accumulated_gyro_error[i] += (sensors_status_imu.gyro_inconsistency_rad_s[i] - angle_rate_threshold) * time_step_s;
```

Comment:
Should this be limited to gyros the sensors module has deemed healthy (no timeouts, etc)?

Reply:
Are you talking about an error that would change primary instance? In such an event, then the accumulated error scores may need to be reset, depending on how the ekf selector handles the change in primary instance. I need to discuss this with you on a call.